src/core/queue/BaseQueue.ts

```diff
@@ -0,0 +1,158 @@
+      this.processJobs();
+    } catch (e) {
+      featLog('error on addJob', e);
+    }
+  }
+
+  addPriorityJob(job: IPriorityQueueJob) {
+    if (!job) throw new Error('Job is required');
+    if (!job?.task) throw new Error('Task is required');
+    if (!job?.priority) throw new Error('Priority is required');
+    if (!job?.operationLabel) throw new Error('Database Operation Label is required');
+    if (!job?.id) throw new Error('id is required');
+
+    job.label = `${job.operationLabel}-${job.id}`;
+
+    /**
+     * Timekeeping block (START)
+     *
+     * This block is to optimize priority job addition to the queue
+     * 1. Find job that has the same label on the timekeeper hash map. If no job with the same label set default lastAddedTime to 0;
+     * 2. If there are jobs with the same label, check its time difference with the timekeeper.
+     * 3. If the difference is more than 50ms, add it to the queue.
+     * 4. forceAddToQueue parameter ignores all above requirement.
+     * 5. Set timekeeper for job with current timestamp.
+     */
+    const currentTime = Date.now().valueOf();
+    const lastAddedTime = this.timekeeper?.[job.label] || 0;
+    const diff = currentTime - lastAddedTime;
+
+    const shouldQueueAdded = lastAddedTime === 0 || diff > 50 || job.forceAddToQueue;
+
+    if (shouldQueueAdded) {
+      this.timekeeper[job.label] = currentTime;
+      this.highPriorityJobs.push({
+        ...job,
+        createdAt: currentTime,
+        label: `${job.operationLabel}-${job.id}`
+      });
+    }
+
+    /**
+     * Timekeeping block (END)
+     */
+
+    this.processJobs();
+  }
+
+  async processJobs() {
+    if (this.isExecutingJob) return;
+
+    featLog(
+      'Remaining High Jobs:',
+      this.highPriorityJobs.length,
+      'Regular jobs:',
+      this.jobs.length
+    );
+
+    this.isExecutingJob = true;
+    if (this.highPriorityJobs.length > 0) {
+      this.highPriorityJobs.sort(
+        (a, b) => a.priority - b.priority || (a?.createdAt || 0) - (b?.createdAt || 0)
+      );
+      const job = this.highPriorityJobs.at(0);
+      if (job) {
+        featLog('Processing High Priority Job:', job.label, 'Priority:', job.priority);
+        const response = await job.task();
+        this.highPriorityJobs.shift();
+        if (job?.callback) job.callback(response);
+      }
+    } else if (this.jobs.length > 0) {
+      const job = this.jobs.at(0);
+      if (job) {
+        featLog('Regular jobs:', job.label);
+        const response = await job.task();
+        this.jobs.shift();
+        if (job?.callback) job.callback(response);
+      }
+    }
+
+    this.isExecutingJob = false;
+
+    if (this.highPriorityJobs.length > 0 || this.jobs.length > 0) {
+      this.processJobs();
+    } else {
+      featLog('No more jobs to process');
+      this.timekeeper = {};
+    }
+  }
+
+  getRemainingJobsCount() {
+    const jobsCount = this.jobs.length + this.highPriorityJobs.length;
+    return jobsCount;
+  }
+}
```

Comment:
<img src="https://avatars.githubusercontent.com/in/347564?s=41" alt="Image description" width="20" height="20">   CodeRabbit

The `BaseQueue` class implementation looks good overall. It provides a way to manage jobs with different priorities and ensures that high-priority jobs are processed first. The use of a singleton pattern for the queue instance is also a good practice.

However, there's a potential issue in the `processJobs` method. If an error occurs during the execution of a job task (either regular or high priority), it will cause the entire process to stop and no further jobs will be processed until a new job is added. This could lead to unprocessed jobs remaining in the queue indefinitely if they continually throw errors.

To handle this, you should add error handling within the `processJobs` method to catch any exceptions thrown by the job tasks. This way, even if a job fails, the queue can continue processing the remaining jobs. Here's a suggested change:

```diff
  async processJobs() {
    if (this.isExecutingJob) return;

    featLog(
      'Remaining High Jobs:',
      this.highPriorityJobs.length,
      'Regular jobs:',
      this.jobs.length
    );

    this.isExecutingJob = true;
    if (this.highPriorityJobs.length > 0) {
      this.highPriorityJobs.sort(
        (a, b) => a.priority - b.priority || (a?.createdAt || 0) - (b?.createdAt || 0)
      );
      const job = this.highPriorityJobs.at(0);
      if (job) {
        featLog('Processing High Priority Job:', job.label, 'Priority:', job.priority);
        try {
          const response = await job.task();
          this.highPriorityJobs.shift();
          if (job?.callback) job.callback(response);
        } catch (e) {
          featLog('error on processJobs - highPriorityJobs', e);
          this.highPriorityJobs.shift();
        }
      }
    } else if (this.jobs.length > 0) {
      const job = this.jobs.at(0);
      if (job) {
        featLog('Regular jobs:', job.label);
        try {
          const response = await job.task();
          this.jobs.shift();
          if (job?.callback) job.callback(response);
        } catch (e) {
          featLog('error on processJobs - jobs', e);
          this.jobs.shift();
        }
      }
    }

    this.isExecutingJob = false;

    if (this.highPriorityJobs.length > 0 || this.jobs.length > 0) {
      this.processJobs();
    } else {
      featLog('No more jobs to process');
      this.timekeeper = {};
    }
  }
```

This change adds a `try-catch` block around the execution of each job task. If an error is thrown, it's logged and the job is removed from the queue, allowing the next job to be processed.


<!-- This is an auto-generated comment by OSS CodeRabbit -->